src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -497,6 +499,23 @@ public static int Count<T>(this ImmutableArray<T> items, Func<T, bool> predicate
             return count;
         }
 
+        [StructLayout(LayoutKind.Sequential)]
+        private struct ImmutableArrayProxy<T>
+        {
+            internal T[] MutableArray;
+        }
+
+        // TODO(https://github.com/dotnet/corefx/issues/34126): Remove when System.Collections.Immutable
+        // provides a Span API
+        internal static T[] DangerousGetUnderlyingArray<T>(this ImmutableArray<T> array)
+            => Unsafe.As<ImmutableArray<T>, ImmutableArrayProxy<T>>(ref array).MutableArray;
```

Comment:
Looks like `As<ImmutableArray<T>, T[]>` works as well?

Reply:
Ehhh that scares me, because what if there's padding? I think Sequential layout should at least guarantee that if padding is added, it's the same amount.

Reply:
Would would a reference type require padding in all circumstances? Since we're relying on the fact that there's only a single array anyways, I think we're fine here, no?

Reply:
That being said, since this is a hard guarantee in any case I think it's the right way to do it. +1